[그리디] 김지우 Spring Core 8,9,10 단계 제출합니다.

Ji-Woo-Kim opened the pull request:
승현님, 리뷰로 처음 뵙게 되어 반갑습니다!
잘 부탁드립니다 🙇‍♀️ 

스프링 미션으로 처음 뵙다보니 제가 어느 수준인지 데이터가 없으실 것 같아 간략하게 말씀드리자면,
스프링 프로젝트 경험은 있으나 자바도 잘 모르는 상태에서 하게 된 프로젝트라 주도적으로 기능을 구현하고 참여하기 보단 팀원들의 코드를 보며 학습하는 과정에 의의를 두었던 경험이 있습니다! Spring에 대한 제대로된 학습과정을 현재 그리디에서 시작하는 것으로 봐주시면 됩니다 😃 

## 미션 진행과정
기존 일정으론 원래 이번 주까지 JDBC를 마치고
다음 주부터 Core를 시작하기로 예정되어 있었지만,
이번 주까지 Core, 즉 스프링 기초 마지막 과정을 
끝내는 것으로 정정되었습니다.

그래서 맨 처음 커밋이 Spring JDBC 7단계 관한 커밋인데요,
사실 main 로직이 변경된 것은 없고 단지 nextstep에서 제공하는 칠단계() 테스트 메서드만 추가하여 잘 돌아가는지 확인하는 것으로 JDBC는 마무리 했습니다!

### 8️⃣ 8단계 
: 시간 관리 기능

💬 처음 예약관리 기능을 구현한 던 것을 참고하여 이전 단계들을 복습하는 느낌으로 진행했습니다.
- 시간 테이블, 시간 객체 구현
- Reservation과비슷한 방식으로 TimeDao, TimeService, TimeController 구현
- 팔단계() 테스트 통과 확인

### 9️⃣ 9단계 
: 예약 기능에서 시간을 시간 테이블에 저장된 값만 선택할 수 있도록 기존 코드 수정

💬 테이블을 조인하여 로직을 구현하는 것은 처음 해보았습니다. next-step에서 제공되는 `세부 요구사항`을 따라 차근차근하니 큰 무리는 없었지만 추가적으로 학습하거나 고민할 포인트가 있다면 말씀해주시기 바랍니다!
- 테이블 스키마 재정의 : 외래키 지정
- domain.Reservation : Time 객체를 필드로 갖도록 수정
- ReservationDao : innerjoin 이용한 쿼리로 수정
- 구단계() 테스트 통과 확인

이때, `localhost:8080`으로 접속하여 확인할 때 예약 등록이 안 되는 오류가 있었습니다. (<- `확인` 버튼을 눌러 등록 요청하는 기능이 아예 먹히지 않음)
자바스크립트 코드와 서버 요청dto 필드 이름이 일치않아 생겼던 문제로 확인되어 nextstep에서 제공하는 `new-reservation.js` 코드를 `time -> timeId`로 수정하였습니다.

### 🔟 10단계
: 레이어드 아키텍처를 적용하여 레이어별 책임과 역할에 따라 클래스 분리

💬 10단계에 도달하기 전부터 계층을 분리하여 작업해왔기 때문에 사실 별다른 작업을 하지 않았는데요!
 조금 고민을 해왔던 부분이라면 controller입니다. 
현재 home/reservation/time 이렇게 도메인위주로 분리하여 관리 중인데요, 
페이지를 랜더링하는(즉 뷰만 반환하는) controller와 
Json을 반환하며 CRUD 기능을 포함한 controller로 
분리하는 것이 좋을지에 대한 고민이 있었습니다.
이전 단계에서 다른 리뷰어님들과도 이야기를 나누어봤는데, 
일반 controller와 rest한 controller로 나누는 것이 좋겠다는 의견도 있으셨지만 
이렇게 되는 경우 페이지를 랜더링하는 컨트롤러가 나중에 너무 무거워질 수도 있다는 의견도 있으셨습니다. 
그래서 승현님은 의견은 어떠신지 궁금하네요 😅 


---
시간 여유 있으실 때 천천히 리뷰 부탁드려요~!
감사합니다!

Ji-Woo-Kim committed:
test: 칠단계() 테스트 추가
- ReservationController create() Location 헤더 추가

Ji-Woo-Kim committed:
style: 코드 스타일 일관성 유지 및 개행 추가

Ji-Woo-Kim committed:
feat: 시간 테이블, 객체 구현

Ji-Woo-Kim committed:
feat: 시간 요청,응답 DTO 구현

Ji-Woo-Kim committed:
feat: TimeDao 구현

Ji-Woo-Kim committed:
feat: TimeService 구현
- saveTime(): 시간 등록
- findAllTimes(): 등록된 모든 시간 조회
- deleteTime(): 등록된 시간 삭제

Ji-Woo-Kim committed:
feat: TimeController 구현
- 시간관리 페이지 랜더링
- 시간 등록 api
- 등록된 시간 조회 api
- 시간 삭제 api

Ji-Woo-Kim committed:
test: 팔단계() 테스트 메서드 추가
- 시간 관리 기능 테스트

Ji-Woo-Kim committed:
refactor: new-reservation 페이지 랜더링으로 수정

Ji-Woo-Kim committed:
refactor: 외래키 지정

Ji-Woo-Kim committed:
refactor: 시간 타입을 Time객체로 수정

Ji-Woo-Kim committed:
refactor: Reservation&Time 객체 조인, 쿼리 수정

Ji-Woo-Kim committed:
test: 조인 된 ReservationDao CRUD 테스트

Ji-Woo-Kim committed:
refactor: Reservation&Time 조인 적용

Ji-Woo-Kim committed:
test: 구단계() 테스트 추가

Ji-Woo-Kim committed:
fix: 요청DTO 필드와 맞지 않던 time → timeId로 수정하여 예약 등록 오류 해결

BackFoxx commented on src/main/java/roomescape/dao/ReservationDao.java:
```diff
@@ -47,24 +54,44 @@ public Reservation addReservation(ReservationRequestDto dto) {
     }
```
reservation을 저장해서 등록된 id를 가져왔다면 인자로 넘어온 Reservation에 id만 합쳐 반환하면 되었을텐데
findById로 DB에서 다시 조회해 가져오는 이유가 있으신가요?

Ji-Woo-Kim replied:
insert 이후에 `findById()`로 다시 조회한 이유는
단순히 ID만 합쳐 반환할 경우 DB에 실제로 저장된 값이 반영되지 않을 수 있기 때문입니다. (= DB와의 싱크가 맞지 않을 수 있다는 우려)
특히 현재 구조에서는 `Reservation` 안에 `Time` 객체가 포함되어 있고, 
`findById()` 내부에서 Time과 조인한 쿼리를 통해 재조회하고 있는데요.
이 방식을 통해 완성된 Reservation 객체를 반환하는 것이 
데이터에 대해 더 안정성이 보장된다고 생각했습니다.

그치만, 성능을 고려한다면 ID만 세팅하는 방식을 택하는 것이 
더 좋을 것 같습니다.!

그럼 현재 구조를 다음처럼 수정할 수 있을 것 같은데요,
1. insert 후 생성된 ID를 기존 Reservation 객체에 `setId()`하여 바로 반환
2. insert 전에 `timeDao.findById()`로 시간 정보 미리 조회한 뒤, Reservation에 `setTime()`
3. Join 없이 두 테이블 분리 (단, 미션 요구사항에 `Reservation`에 `Time` 객체를 필드로 포함하라는 조건이 있었습니다)

성능과 데이터 완결성 사이에서 tradeoff가 있는 부분이라 고민이 길어졌는데요, 위 방식 중 2번 방식을 택하여 리팩토링 해보았습니다.

Ji-Woo-Kim replied:
Dao 코드, 조인한 쿼리 이용 등 처음 학습하게 된 부분이 있어 다른 코드를 참조하여 작성하였습니다. 코멘트 받고 난 뒤, DB 재조회하는 로직에 대해 다시 생각해보았는데요.!

현재 구조에서는 `Reservation` 안에 `Time` 객체가 포함되어 있고, 
`findById()` 내부에서 Time과 조인한 쿼리를 통해 재조회하고 있습니다.
그러나, 단순히 ID만 합쳐 반환할 수도 있는데 이런 경우 DB에 실제로 저장된 값이 코드에 반영되지 않을 수 있을 것 같습니다. (DB와의 싱크가 맞지 않을 문제)
따라서 재조회하여 reservation객체를 반환하는 것이 데이터의 안정성이 더 보장된다고 생각됩니다.

그치만, DB를 재조회하는 만큼 후에 데이터가 늘어날수록 성능이 떨어진다는 문제가 또 있을 수 있겠네요..
성능과 데이터 완결성 사이에서 tradeoff가 있는 부분이라 생각됩니다..!

banlife-shjo replied:
멋진 고민을 해주셨군요 :) 
DB에 실제로 저장된 값이 코드에 반영되지 않는 경우로 어떤 게 있을까요?

BackFoxx commented on src/main/java/roomescape/service/TimeService.java:
```diff
@@ -0,0 +1,36 @@
+    public TimeResponseDto saveTime(TimeRequestDto requestDto) {
+        Time time = new Time();
+        time.setTime(requestDto.getTime());
+        Time saved = timeDao.save(time);
```
동일한 시각을 중복으로 저장할 수도 있나요?

Ji-Woo-Kim replied:
Time 테이블을 10:00, 11:00, 12:00... 같은 고정된 시간을 저장해두는 마스터 테이블이라 두는 상황이기에 같은 10:00을 두 번 저장할 이유가 없는데 이 부분에 대한 방어로직이 전혀 없었네요 😢 

1. schema.sql에서 time_value에 대해 UNIQUE 조건 추가
2. TimeDao에서 `existsByTime()`로 DB 조회하여 중복 카운트 유무 확인
3. TimeService에서 중복 메서드 호출로 확인 후 예외 발생

이렇게 방어로직 추가해두었습니다!

banlife-shjo replied:
완벽해요 !

BackFoxx commented on src/main/java/roomescape/service/TimeService.java:
```diff
@@ -0,0 +1,36 @@
+
+    public TimeResponseDto saveTime(TimeRequestDto requestDto) {
+        Time time = new Time();
+        time.setTime(requestDto.getTime());
```
생성자 대신 setter를 통해 시간 정보를 주입한 이유가 있으실까요?

Ji-Woo-Kim replied:
아, 이 부분은 간단히 요청 값을 기준으로 객체를 생성한다는 생각에 setter로 값을 주입했습니다.

코멘트 보고 나니, 생성자에서 값 주입을 한 번에 처리하는 방식이 더 안전할 수 있을 것 같네요.!
요청dto에서 값을 가져와서 넣어주는 방식으로 리팩토링해두었습니다.

banlife-shjo replied:
setter보다 한 단계 나아진 것은 맞지만
모든 필드에 대한 생성자인 AllArgsConstructor를 사용하다보니 id를 초기화할 필요가 없는 경우에도 `new Time(null, requestDto.getTimeValue())`처럼 사용해야 한다는 점이 불편해 보입니다.

필요한 필드만 선택적으로 초기화하면서도 객체 생성 코드를 가독성있게 만들어주는 @Builder라는 어노테이션을 사용해 보시겠어요?
lombok에서 제공하는 기능인데, 다양한 활용 방법이 있으니 쭈루룩 학습해보시면 좋을 것 같아요.

BackFoxx commented on src/main/java/roomescape/domain/Reservation.java:
```diff
@@ -18,5 +16,5 @@ public class Reservation {
     private String name;
     private LocalDate date;
-    private LocalTime time;
+    private Time time;
```
좋은 시도입니다.
그런데 외래키로 연결된 시간 정보가 12시에서 10시로 변경되거나 데이터가 삭제되면 예약정보는 어떻게 되나요?

Ji-Woo-Kim replied:
현재 구조에서는 `Reservation`이 `Time`을 참조하고 있기 때문에,
Time 값이 수정되거나 삭제되면 예약 정보에도 영향을 주는데 이에 대한 방어 설계가 안 되어 있는 점이 아쉽네요 🥲 

해당 시간을 참조 중인 예약 건이 하나라도 있다면 예외를 던져서 
시간 정보 삭제/수정을 제한하는 로직을 추가해두도록 하겠습니다.

그러나, 이 방식이 데이터양이 많아질수록 성능 이슈가 있을 것으로 예상됩니다.
혹시 다른 추천하시는 방안이 있으실까요?
아님 서비스 단에 예외처리하는 것으로도 괜찮은 방안일까요?

Ji-Woo-Kim replied:
아.. 그러나 이 방식으로 한 후 테스트 해보니 프론트코드와 DB 연동 문제가 있는 것 같습니다.
시간을 삭제하려고 할 때 서버에서 예외는 잘 나타는데, 
이후 예약 관리 페이지에서 목록이 깨지게 되는 문제가 있습니다...

banlife-shjo replied:
```
해당 시간을 참조 중인 예약 건이 하나라도 있다면 예외를 던져서
시간 정보 삭제/수정을 제한하는 로직을 추가해두도록 하겠습니다.
```

성능 이슈 이전에 애플리케이션이 이런 정책을 가지고 있으면 운영하기 힘든 서비스가 만들어집니다.
한강 오리배 타기 상품이 있고 오전 10시, 오전 11시 옵션이 있다고 상상해 봅시다.
10시 옵션에 예약이 8개 있었는데, 비 때문에 10시 옵션을 닫아야 합니다.
그런데 10시 옵션으로 이미 들어온 예약이 있어서 옵션을 지울 수도 없고 다른 시간으로 바꿀 수도 없습니다.
오리배 사장님이 화딱지 나지 않으실까요?

banlife-shjo replied:
제가 보기에는 방어 로직 이전에 Reservation이 Time 정보를 '참조' 형태로 가지고 있어도 되는지 먼저 고민해보시면 좋을 것 같아요!

BackFoxx reviewed (changes requested):
안녕하세요! 상세하게 PR을 작성해주신 덕에 수월하게 리뷰할 수 있었습니다.
정성스럽게 미션을 진행해주셔서 감사합니다.

---

저는 아무래도 회사 코드를 다루다보니 회사 컨벤션에 맞게 코드를 짜는데,
컨트롤러 또한 'uri의 prefix를 기준으로 나눈다'는 규칙에 따라 나누고 있습니다.
회사에서도 여러 고민 끝에 채택한 컨벤션인 만큼 그 이유도 나름 타당한데,
다른 컨벤션들을 채택했을 때의 문제점들을 대체로 무난하게 커버하기 때문입니다.
* **일반 controller와 rest한 controller로 나누는 경우** : 약간 잘못된 문장인데, html을 응답한다고 해서 rest하지 않은 건 아닙니다. 정확히는 content-type별로 컨트롤러를 나누는 것에 가까운데, 이 경우 html이나 json 외에 다른 포맷을 응답하는 소수의 케이스를 위해 ReservationTsvController, ReservationCsvController처럼 별도의 컨트롤러를 또 만들어야 할 수 있습니다.
* **도메인 별로 컨트롤러를 나누는 경우** : 도메인 간의 구분이 엄격한 프로젝트에서는 해당 컨벤션이 유효합니다. 그러나 네이버의 메인 페이지처럼 여러 도메인을 짬뽕때린 응답을 내려주어야 하는 경우 '이 API는 어느 도메인이지'라는 고민을 할 수밖에 없습니다.

---

이번 리뷰는 비즈니스 로직의 유효성 위주로 리뷰를 남겨드렸어요!
코멘트로 남긴 문제점을 해결하는 방법으로 여러 가지가 있는데, 여러 방법을 고민해보시고 최선의 해결방법을 적용해보세요. 그리고 시간이 나신다면 어떤 고민을 하셨는지도 글로 함께 남겨주세요.

Ji-Woo-Kim committed:
feat: 동일한 시각 중복 저장 방지 로직 추가
- schema.sql: time_value UNIQUE 조건 추가
- existsByTime(): 중복 유무 확인
- 중복 시 예외 발생

Ji-Woo-Kim committed:
style: 코드 일관성과 가독성 위한 수정
- 어노테이션 수정
- 변수명 수정

Ji-Woo-Kim committed:
style: 코드 일관성 - 변수명 수정

Ji-Woo-Kim committed:
feat: Time 삭제 관련 예외처리 로직 구현
- ON DELETE/UPDATE RESTRICT 제약 추가
- ReservationDao: existsByTimeId()
- TimeService에서 삭제 전에 메서드 호출하여 검사

Ji-Woo-Kim committed:
fix: 필드명 통일
- time 네이밍을 timeValue로 변경하며 생긴 연동 문제 해결

Ji-Woo-Kim committed:
refactor: save() setter -> 생성자 주입
